### PR TITLE
TYP: some types for pandas/core/arrays/sparse/array.py

### PR DESCRIPTION
**Pull request by simonjayhawkins**
broken off #28339

### CODE REVIEW

File: pandas/core/arrays/sparse/array.py [hunk 1]
```diff
@@ -50,6 +50,8 @@
 
 from .dtype import SparseDtype
 
+_SparseArrayT = TypeVar("_SparseArrayT", bound="SparseArray")
```
Comment (simonjayhawkins) [inline pandas/core/arrays/sparse/array.py]: @jreback do you want this public

File: pandas/core/arrays/sparse/array.py [hunk 2]
```diff
@@ -372,8 +375,11 @@ def __init__(
 
     @classmethod
     def _simple_new(
-        cls, sparse_array: np.ndarray, sparse_index: SparseIndex, dtype: SparseDtype
-    ) -> ABCSparseArray:
+        cls: Type[_SparseArrayT],
```
Comment (jbrockmendel) [inline pandas/core/arrays/sparse/array.py]: needing to annotate `cls` seems weird.  is that just me?

File: pandas/core/arrays/sparse/array.py [hunk 3]
```diff
@@ -805,13 +811,14 @@ def _get_val_at(self, loc):
...
 
-    def take(self, indices, allow_fill=False, fill_value=None):
+    def take(self, indices, allow_fill: bool = False, fill_value=None):
         if is_scalar(indices):
             raise ValueError(
                 "'indices' must be an array, not a scalar '{}'.".format(indices)
             )
         indices = np.asarray(indices, dtype=np.int32)
 
+        result: List
```
Comment (jbrockmendel) [inline pandas/core/arrays/sparse/array.py]: not saying anything about contents?

File: pandas/core/arrays/sparse/array.py [hunk 4]
```diff
@@ -1392,8 +1399,10 @@ def __abs__(self):
     # ------------------------------------------------------------------------
 
     @classmethod
-    def _create_unary_method(cls, op):
-        def sparse_unary_method(self):
+    def _create_unary_method(
+        cls: Type[_SparseArrayT], op
+    ) -> Callable[[_SparseArrayT], _SparseArrayT]:
```
Comment (jbrockmendel) [inline pandas/core/arrays/sparse/array.py]: does mypy specifically need this?  my kneejerk reaction is that it adds verbosity but not clarity

**Comment by jbrockmendel** [comment]
small comments, non-blockers, LGTM

File: pandas/core/arrays/sparse/array.py (hunk 4 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/sparse/array.py]: we could see how far we get without it. but remember your frustration with subclass of ChainMap not returning the same type?

File: pandas/core/arrays/sparse/array.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/sparse/array.py]: annotated for generic self. see https://mypy.readthedocs.io/en/latest/generics.html#generic-methods-and-generic-self

again could see how far we get without this.

File: pandas/core/arrays/sparse/array.py (hunk 3 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/sparse/array.py]: not until _take_with_fill and _take_without_fill are annotated. removed till then.

File: pandas/core/arrays/sparse/array.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/sparse/array.py]: no longer applicable

File: pandas/core/arrays/sparse/array.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/arrays/sparse/array.py]: In case I haven't mentioned it recently, I'm totally happy to defer to you when it comes to annotations

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks @simonjayhawkins

**Pull request merged**